Meadowlark-js/packages/meadowlark-authz-server/src/security/TokenValidator.ts

```diff
@@ -0,0 +1,134 @@
+      },
+    };
+  }
+
+  if (jwtStatus.isValid) {
+    return { isValid: true, roles: jwtStatus.roles, clientId: jwtStatus.clientId ?? '' };
+  }
+
+  return {
+    isValid: false,
+    errorResponse: {
+      statusCode: 401,
+      body: '{ "error": "invalid_token", "error_description": "Invalid authorization token" }',
+      headers: { 'WWW-Authenticate': 'Bearer' },
+    },
+  };
+}
+
+export function validateAdminTokenForAccess(authorizationHeader: string | undefined): AuthorizationResponse | undefined {
+  const validateTokenResult: ValidateTokenResult = validateTokenForAccess(authorizationHeader);
+
+  if (!validateTokenResult.isValid) return validateTokenResult.errorResponse;
+
+  if (!hasAdminRole(validateTokenResult.roles)) {
+    return {
+      statusCode: 403,
+      body: '',
+      headers: { 'WWW-Authenticate': 'Bearer' },
+    };
+  }
+
+  return undefined;
+}
+
+/**
+ * Returns true if the client id is found in the authentication store
+ */
+async function isValidClientId(clientId: string | undefined, traceId: string): Promise<boolean> {
+  if (clientId == null) return false;
+
+  // Go to authentication datastore
+  const result: GetAuthorizationClientResult = await getAuthorizationStore().getAuthorizationClient({
+    clientId,
+    traceId,
+  });
+
+  return result.response === 'GET_SUCCESS';
+}
+
+export type IntrospectionResponse = { isValid: false } | { isValid: true; introspectedToken: IntrospectedToken };
+
+/*
+ * Verifies that a bearer token is valid, not expired, etc. Returns parsed claim data
+ * as a ParsedBearerToken. Uses nJWT verify()
+ */
+export async function introspectBearerToken(bearerToken: string, traceId: string): Promise<IntrospectionResponse> {
+  Logger.debug('TokenValidator.introspectBearerToken token', traceId, bearerToken);
+
+  try {
+    const verified: nJwt | undefined = verify(bearerToken, signingKey());
+    if (verified == null) return { isValid: false };
+    const jwt: Jwt = verified as Jwt;
+
+    const nJwtErrorMessages = ['Signature verification failed', 'Jwt cannot be parsed'];
+
+    if (jwt == null || nJwtErrorMessages.includes(jwt.message)) return { isValid: false };
+
+    const isNotExpired: boolean = jwt.message !== 'Jwt is expired';
+    const isMeadowlarkToken: boolean = jwt.body?.iss === TOKEN_ISSUER && jwt.body?.aud === TOKEN_ISSUER;
+
+    const active: boolean = isNotExpired && isMeadowlarkToken && (await isValidClientId(jwt.body?.client_id, traceId));
+
+    return {
+      isValid: true,
+      introspectedToken: {
+        active,
+        client_id: jwt.body?.client_id ?? '',
+        sub: jwt.body?.sub ?? '',
+        aud: jwt.body?.aud ?? '',
+        iss: jwt.body?.iss ?? '',
+        exp: jwt.body?.exp ?? 0,
+        iat: jwt.body?.iat ?? 0,
+        roles: jwt.body?.roles ?? [],
+      },
+    };
+  } catch (err) {
+    Logger.debug('TokenValidator.introspectBearerToken user-submitted token error', traceId, err);
```

Comment:
Are you sure that such an exception is just a user problem, not a problem with our software?

Reply:
Actually, it's neither! Turns out nJwt throws exceptions for not-very-exceptional conditions, like expired token. Fixed this.